### PR TITLE
Fix for duplicate vtable field names

### LINKED ISSUE #30: Codegen issue: vtable struct with duplicate components
This test requires an additional check in `struct_type()` of `cbmc/goto_program/typ.rs` to ensure that the struct's components are unique. In this case, the vtable contains two `foo` fn pointers.

```
trait A {
    fn foo(&self) -> i32;
}

trait B {
    fn foo(&self) -> i32;
}

trait T : A + B {
}

struct S {
    x : i32,
    y : i32,
}

impl S {
    fn new(a:i32, b:i32) -> S {
        S {x:a, y:b}
    }
    fn new_box(a:i32, b:i32) -> Box<dyn T> { 
        Box::new(S::new(a,b))
    }
}


impl A for S {
    fn foo(&self) -> i32{
        self.x
    }
}

impl B for S {
    fn foo(&self) -> i32{
        self.y
    }
}

impl T for S {
}

fn main() {
    let t = S::new_box(1,2);
    let a = <dyn T as A>::foo(&*t);
    assert!(a == 1);
    let b = <dyn T as B>::foo(&*t);
    assert!(b == 2);
}
```

### PR DESCRIPTION
**Pull request by avanhatt**
### Description of changes: 

This is a fix for https://github.com/model-checking/rmc/issues/30, for now keeping the string-matching based vtable implementation. I'm tracking a bigger refactor overhaul in https://github.com/model-checking/rmc/issues/187.

The root cause of #30 was that for vtable field names, we were using just the function name (`foo`), instead of the function name scoped to the specific trait (`A::foo` vs `B::foo`). In addition, I've added a new duplicate trait test that failed under the previous implementation. The fix there is to build the dynamic pretty name by first grabbing the instance and using that, which let me remove the split pretty name function I added in eb6cb24e1db7f930b038ec79cf182b7b0eb4b79e.  

There is a distinction in two different names that are important for vtables:
1. The name for the function that is stored in the vtable itself. One vtable exists per trait + object combination. Because the trait for a vtable can be a sum trait, functions might have the same name, so we need to scope by trait (`A::foo`).
2. The name for the function to be stored in the symbol table for the translation to C. This needs to be scoped globally (`crate::A::foo`).

This fix mostly is about #1, though a change to #2 was necessary to correct new tests.

### Resolved issues:

Resolves #30

### Call-outs:

String matching is brittle, #187 is the end goal.

### Testing:

* How is this change tested?
Existing regression tests plus 4 new ones.

* Is this a refactor change?
No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs [hunk 1]
```diff
@@ -671,15 +671,23 @@ impl<'tcx> GotocCtx<'tcx> {
         trait_ref_t: Binder<'_, TraitRef<'tcx>>,
         t: Ty<'tcx>,
     ) -> Expr {
-        let function_name = self.tcx.item_name(def_id).to_string();
+        let vtable_field_name = self.vtable_fild_name(def_id);
```
Comment (avanhatt) [inline compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs:674]: Oops, I can't spell, will fix

File: compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs [hunk 2]
```diff
@@ -671,15 +671,23 @@ impl<'tcx> GotocCtx<'tcx> {
...
-        let function_name = self.tcx.item_name(def_id).to_string();
+        let vtable_field_name = self.vtable_field_name(def_id);
         let vtable_type_name = aggr_name(&self.vtable_name(t));
         let field_type = self
             .symbol_table
-            .lookup_field_type(&vtable_type_name, &function_name)
+            .lookup_field_type(&vtable_type_name, &vtable_field_name)
             .cloned()
             .unwrap();
+        let instance = Instance::resolve(
```
Comment (danielsn) [inline compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs:681]: Explain why this change in a comment

File: compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/typ.rs [hunk 3]
```diff
@@ -653,13 +653,11 @@ impl Type {
...
     /// }
     pub fn struct_type(name: &str, components: Vec<DatatypeComponent>) -> Self {
-        // TODO: turn this on after fixing issue #30
-        // <https://github.com/model-checking/rmc/issues/30>
-        //assert!(
-        //    Type::components_are_unique(&components),
-        //    "Components contain duplicates: {:?}",
-        //    components
-        //);
+        assert!(
```
Comment (danielsn) [inline compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/typ.rs:656]: nice

File: compiler/rustc_codegen_llvm/src/gotoc/metadata.rs [hunk 4]
```diff
@@ -199,22 +194,14 @@ impl<'tcx> GotocCtx<'tcx> {
...
-        trait_ref_t: Binder<'_, TraitRef<'tcx>>,
-    ) -> String {
-        let normalized_object_type_name = self.normalized_name_of_dynamic_object_type(trait_ref_t);
-        format!(
-            "{}::{}::{}",
-            self.tcx.crate_name(def_id.krate),
-            normalized_object_type_name,
-            self.tcx.item_name(def_id)
-        )
+    /// For the vtable field name, we need exactly the dyn trait name and the function
```
Comment (danielsn) [inline compiler/rustc_codegen_llvm/src/gotoc/metadata.rs:197]: I really like that this is a unified call

File: compiler/rustc_codegen_llvm/src/gotoc/metadata.rs [hunk 5]
```diff
@@ -199,22 +194,14 @@ impl<'tcx> GotocCtx<'tcx> {
...
+    /// name. The table itself already is scoped by the object type.
+    ///     Example: ::Shape::vol
+    /// Note: this is _not_ the same name for top-level entry into the symbol table,
+    /// which does need more crate and type information. For now, the symbol table
+    /// name is from the pretty_name_from_instance function above.
+    pub fn vtable_field_name(&self, def_id: DefId) -> String {
```
Comment (danielsn) [inline compiler/rustc_codegen_llvm/src/gotoc/metadata.rs:203]: comment explaining why we chose this particular printer, and why

**Review by danielsn — COMMENTED** [review]
Do we think this will work across crates?

**Comment by avanhatt** [comment]
> Do we think this will work across crates?

I think so, added tests for `different_crates.rs` and `different_crates_fail.rs` to try and exercise this.

File: compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs (hunk 2 above)
Comment (avanhatt) [reply compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs:685]: Added!

File: compiler/rustc_codegen_llvm/src/gotoc/metadata.rs (hunk 5 above)
Comment (avanhatt) [reply compiler/rustc_codegen_llvm/src/gotoc/metadata.rs:203]: Done

**Review by danielsn — APPROVED** [review]

**Pull request merged**